extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/DataSourcePlan.java

```diff
@@ -137,6 +139,30 @@ public static DataSourcePlan forDataSource(
     } else if (dataSource instanceof LookupDataSource) {
       checkQuerySegmentSpecIsEternity(dataSource, querySegmentSpec);
       return forLookup((LookupDataSource) dataSource, broadcast);
+    } else if (dataSource instanceof FilteredDataSource) {
+      checkQuerySegmentSpecIsEternity(dataSource, querySegmentSpec);
+      return forFilteredDataSource(
+          queryKit,
+          queryId,
+          queryContext,
+          (FilteredDataSource) dataSource,
+          querySegmentSpec,
+          maxWorkerCount,
+          minStageNumber,
+          broadcast
+      );
+    } else if (dataSource instanceof UnnestDataSource) {
+      checkQuerySegmentSpecIsEternity(dataSource, querySegmentSpec);
```

Comment:
Were you able to confirm if the check is required?

Reply:
I have asked around, I'll update it once I get back the comments.

Reply:
Any updates on this? I wanted to get a final check on this before the approval

Reply:
Based on the UTs I ran, this can be removed for the top-level UnnestDS and FilteredDS, as the inner calls of `forDataSource` on the base data sources will validate this. For example, if the base is a query data source that needs the check, the recursive call will take care of it on the base. Hence I removed this